### PR TITLE
支持等待 .detach() 创建的协程

### PR DESCRIPTION
**Pull request by microcai**
当调用 .detach(), 然后不等待它，那么这个 detch 的协程就直接后台跑了。就好像 create_thread
创建了个新线程一样。

当调用 .detach(), 然后使用 co_await 等待它，那么当前协程就等待这个 detach 的协程完成。
就好像调用 std.thread.join() 那样。

### CODE REVIEW

**Review by Jackarain — COMMENTED** [review]
请简述实现原理

**Review by microcai — COMMENTED** [review]
就是给 awaitable_detached 添加  await_suspend 和 await_resume ， 这样它就能被 co_await 了。

File: include/ucoro/awaitable.hpp [hunk 1]
```diff
@@ -120,31 +120,89 @@ namespace ucoro
...
+
+			void unhandled_exception() { }
+
+			// 对 detached 的 coro 调用 co_await 相当于 thread.join()
+			// 因此记录这个 continuation 为 joiner
+			std::coroutine_handle<> joiner;
+			bool no_owner = false;
+		};
+
+		awaitable_detached(promise_type* promise) : promise(promise) {}
```
Comment (Jackarain) [inline include/ucoro/awaitable.hpp:167]: awaitable_detached 应是明确的构造，即 `explicit`，并注意风格

**Review by Jackarain — COMMENTED** [review]
主要是很多代码风格问题，还有一些功能上的问题：
1，co_await 的情况下，也需要考虑返回值。
2，同样也需要考虑异常转发。

**Comment by microcai** [comment]
OK. 等我发第二版

File: include/ucoro/awaitable.hpp [hunk 2]
```diff
@@ -120,31 +120,89 @@ namespace ucoro
...
+				constexpr bool await_ready() const noexcept { return false; }
 
-			void return_void() noexcept
-			{
-			}
+				std::coroutine_handle<> await_suspend(std::coroutine_handle<>) const noexcept
+				{
+					std::coroutine_handle<void> promise_handle = std::coroutine_handle<promise_type>::from_promise(*this_);
+
+					if (this_->no_owner)
```
[the remark this answers is not in the document]
Comment (microcai) [reply include/ucoro/awaitable.hpp]: 原来如此！学习了！

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -2,6 +2,7 @@
 set_target_properties(ucoro_tests PROPERTIES FOLDER "ucoro_tests")
 
 add_subdirectory(test1)
+add_subdirectory(test2)
```
Comment (Jackarain) [inline tests/CMakeLists.txt:5]: 如果可以的话，我觉得应该将此测试命名为 test_detach 类似更有意义的名称，避免延续我之前一开始的 test1 test2 这类不良的命名风格。

如果有空了，我会将之前的 test1 test2 ... 这些重新规范一下

**Review by Jackarain — COMMENTED** [review]
LGTM，

File: tests/CMakeLists.txt (hunk 3 above)
Comment (microcai) [reply tests/CMakeLists.txt:5]: 那下次统一改？ 还是这次我先改成 test_detach

**Review by Jackarain — APPROVED** [review]
先 Approve，后面再统一修改 tests 的命名问题
